## Tarea 1: Repaso de los conceptos básicos de Python
### Ciencia de datos para físicos: Facultad de ciencias, UNAM
#### Profesores: Karen Rubí Jiménez-López, Pedro Arturo Flores-Silva

**Instrucciones:** Este notebook contiene una serie de ejercicios que deben ser resueltos. Cada ejercicio se encuentra en una celda, ya sea de código o Markdown. Debes crear una o dos *celdas debajo de cada ejercicio* con tu(s) respectiva(s) respuestas. No modifiques las celdas originales.

Recuerda que para ejecutar el código de una celda, presionar `Ctlr + Enter` o bien el boton de `play` en el editor Jupyter notebook.

Cada ejercicio tendrá diferente valor, y se indicará en la celda de éste. La calificación final será la suma de los puntos obtenidos en cada ejercicio. Se calificará considerando la calidad de la respuesta, la claridad de la explicación y la correctitud del código. Por ejemplo, si el código no ejecuta la respuesta correcta pero se da una idea textual, se demuestra que hay comprensión del problema y el código hace sentido para el revisor, se otorgará una calificación parcial. Esto es, las respuestas no solo se evaluarán como bien o mal, sino que se considerará el esfuerzo.

El codigo que realices debes comentarlo donde creas necesario, no vale no hacer ningún comentario. Recuerda que los comentarios son importantes para que el código sea legible y entendible, éstos se agregan con el símbolo #. Para las funciones y/o clases que crees, agrega la documentación correspondiente (docstring), éste se agrega con `""" docstring """`.

### Ejercicio 1 (1 punto):
En una celda Markdown contesta: 
* ¿Qué es una palabra reservada en python? Enlista 5 ejemplos de nombres reservados en python y describe su función.
* Coloca tres ejemplos de nombres de variables que no son válidos en python. Explica por qué no son válidos.

R1=Son palabras las cuales el lenguaje de programación ya ha reservado para realizar ciertas tareas.

R2=Variables como "1X,1_x,class", pues 2 de ellas empiezas con numero y la ultima es una palabra reservada de python.

### Ejercicio 2 (1 punto):

En una celda de código, escribe una función que encuentre las raíces de cualquier polinomio de segundo grado (implementar la chicharronera). En una celda markdown detalla posibles casos en la cual la función podría fallar.
Usa dicha función para encontrar las raíces del polinomio $f(x) = 10x^2 - 2x$ e imprimelas en pantalla.

In [19]:
from math import sqrt#TRAEMOS el operador raiz
#preguntamos variables para poder trabajar
A = int(input("Ingrese el coeficiente de la variable cuadrática\n"))
B = int(input("Ingrese el coeficiente de la variable lineal\n"))
C = int(input("Ingrese el término independiente\n"))
if ((B**2)-4*A*C) < 0:#no trabajaremos con numeros complejos por lo que lo descartamos y ponemos un mensaje
     print("La solución de la ecuación es con números complejos")
else:
    #Ponemos la chicharronera, (+ para x1) y (- para x2)
    x1 = (-B+sqrt(B**2-(4*A*C)))/(2*A)
    x2 = (-B-sqrt(B**2-(4*A*C)))/(2*A)
    print("Las soluciones de la ecuación son:")
    print(x1)
    print(x2)

Ingrese el coeficiente de la variable cuadrática
10
Ingrese el coeficiente de la variable lineal
-2
Ingrese el término independiente
0
Las soluciones de la ecuación son:
0.2
0.0


### Ejercicio 3 (1 punto):

*Caida libre*: El gran Galileo Galilei subió la torre de Pisa para determinar el tiempo que tarda un objeto en caer desde una altura $H$ al suelo. Para ello, colocó un objeto de masa $m=100$ kg en la torre y lo dejó caer. El tiempo que tardó en caer fue $t=0.05616667$ minutos.

Considerando que la aceleración de la gravedad es $g=9.8$ $\frac{m}{s^2}$, escribe un programa que determine la altura de la torre de Pisa. El programa debe imprimir la altura en la pantalla.

Recuerda que el movimiento de caída libre simplificado cumple la ecuación: $y(t) = H + vt - \frac{1}{2}gt^2$.

In [6]:
def altura(t1):#la variable se guardara en t1, 
    #t1=0.056616617
    t=t1*60 #covertimos a segundos
    g=9.8 #valor de la gravedad
    h=(1/2)*g*(t**2) #la funcion y(t) la simplificamos para v0=0 quedando lo que e igualando a cero, 
    # quedando solo h en funcion de la gravedad y el tiempo t
    return h 
print(altura(0.056616617))

56.54398489405553


### Ejercicio 4 (1 punto):
Crea una lista de números enteros del 0 al 99 : `lista = [0,1, 2, 3, 4, ... , 95, 96, 97, 98, 99]` (6 puntos menos si creas la lista manualmente).
 
Escribe un programa que imprima en la pantalla una la lista que cumpla con las siguientes condiciones:
* Los ultimos 10 elementos: debe imprimir lo siguiente `[90, 91, 92, 93, 94, 95, 96, 97, 98, 99]`.
* Los primeros 11 elementos: debe imprimir lo siguiente `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`.
* La serie de elementos de la lista que están entre 60 y 75: debe imprimir lo siguiente `[60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75]`.
* El número 50: debe imprimir lo siguiente `50`.

Debes aplicar el concepto de *slicing* para resolver este ejercicio (notación de puntos `[:]`).

Puntos menos si se crean listas nuevas para cada caso y/o se seleccionan manualmente los valores.

In [7]:
lista=list(range(0,101)) #creamos una lista del 0 al 100. 
print(lista[90:100])#pedimos los 10 ultimos numeros
print(lista[0:11])#pedimos los primeros 11 numeros
print(lista[60:76])#pedimos los numeros del 60 al 75
print(lista[50:51])#pedimos solo el 50

[90, 91, 92, 93, 94, 95, 96, 97, 98, 99]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75]
[50]


### Ejercicio 5 (1 punto):
Crea una lista de números enteros del 0 al 1000. A partir de dicha lista:
* Guarda en una variable una lista con los números pares usando ciclos y condicionales. Imprime los últimos 10 elementos de la lista final.
* Guarda en una variable una lista con los números impares usando listas comprensivas (comprehension list). Imprime los primeros 10 elementos de la lista final.

In [20]:
lista1_1000= list(range(0,1001)) #Mas de lo mismo pedimos una lista del 0 al 1000
def pares(lista1_1000):#aqui va la primera parte de la pregunta
    num_pares = []
    for n in lista1_1000:#traemos los numeros de la lista
        if n % 2 == 0:#pedimos que solo deje los parres
            num_pares.append(n)    
    return num_pares#guardamos la lista pares
num_pares = pares(lista1_1000)
print(lista1_1000)#imprimimos la lista del 1 al 1000 pa que vea que no lo choreo
print(num_pares)#imprimimos la lista de numeros pares pero contiene al 0

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221,

In [21]:
num_pares.pop(0)#quitamos el 0 para que empiece desde 2
#solo se puede ejecutar una vez ya que .pop(0) elimina el primer elemento de la lista, si se vuelve a ejecutar quitara el 2
print(num_pares)#imprimimos la lista para verificaro
print(num_pares[490:500])#imprimimios los ultimos 10 terminos

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208, 210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234, 236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260, 262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286, 288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312, 314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338, 340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 362, 364, 366, 368, 370, 372, 374, 376, 378, 380, 382, 384, 386, 388, 390, 392, 394, 396, 398, 400, 402, 404, 406, 408, 410, 412, 414, 416, 418, 420, 42

In [22]:
num_imp= [n for n in lista1_1000 if n % 2 !=0]#lo mismo pero usando listas comprensivas
print(num_imp)
print(num_imp[490:500])

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 161, 163, 165, 167, 169, 171, 173, 175, 177, 179, 181, 183, 185, 187, 189, 191, 193, 195, 197, 199, 201, 203, 205, 207, 209, 211, 213, 215, 217, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245, 247, 249, 251, 253, 255, 257, 259, 261, 263, 265, 267, 269, 271, 273, 275, 277, 279, 281, 283, 285, 287, 289, 291, 293, 295, 297, 299, 301, 303, 305, 307, 309, 311, 313, 315, 317, 319, 321, 323, 325, 327, 329, 331, 333, 335, 337, 339, 341, 343, 345, 347, 349, 351, 353, 355, 357, 359, 361, 363, 365, 367, 369, 371, 373, 375, 377, 379, 381, 383, 385, 387, 389, 391, 393, 395, 397, 399, 401, 403, 405, 407, 409, 411, 413, 415, 417, 419, 421,

### Ejercicio 6 (2 puntos): Aproximación de $\pi$
Una de las formas más sencillas de aproximar $\pi$ es mediante la siguiente serie infinita de Leibniz:
$$\frac{\pi}{4} = \sum_{n=0}^{\infty}\frac{(-1)^{n}}{(2n+1)} = \frac{1}{1} - \frac{1}{3} + \frac{1}{5}-\frac{1}{7}+\frac{1}{9}-...$$

crea una función para aproximar $\pi$ utilizando la serie de Leibniz. Usa n = 100000 e imprime el valor de tu aproximación tanto el valor de $\frac{\pi}{4}$ como $\pi$.

El valor de $\pi$ se obtiene al despejar:
$$\pi = 4 * \left(\sum_{n=0}^{\infty}\frac{(-1)^{n}}{(2n+1)} \right)$$

La siguiente celda de codigo ofrece el valor de $\frac{\pi}{4}$

In [1]:
import numpy as np
#El valor de pi/4
print("El valor de pi/4 es aproximadamente {}".format(np.pi/4))

El valor de pi/4 es aproximadamente 0.7853981633974483


In [23]:
def serieleibniz(n):
    suma = 0#declaramos la suma para el ciclo for y que se sume n veces
    for i in range(n):
        serie = (-1) ** i /(2*i+1)#escribiendo la expresion matematica
        suma = suma + serie#La suma
    return suma * 4#La suma x4 para encontrar pi
serieleibniz(100000)#n=1000000

3.1415826535897198

### Ejercicio 7 (1 puntos): Identificador de vocales

Escribe una función que reciba una cadena, por ejemplo 'el pErro de la tiA chOnita' y regrese la cantidad de vocales que contiene. La función debe contener varios casos de prueba, if, elif y else. Con la cadena de ejemplo, imprime en pantalla el resultado de la función (10).

Para conocer si un caracter es una vocal, primero debes descomponer la cadena en caracteres individuales. Para ello observa el siguiente código:

In [6]:
# Divide una cadena en caracteres individuales
cadena_prueba = 'el pErro de la tiA chOnita'
caracteres_individuales = list(cadena_prueba)
print(caracteres_individuales)

['e', 'l', ' ', 'p', 'e', 'r', 'r', 'o', ' ', 'd', 'e', ' ', 'l', 'a', ' ', 't', 'i', 'a', ' ', 'c', 'h', 'o', 'n', 'i', 't', 'a']


In [16]:
#tomando el ejemplo de la tarea tenemos
cadena_prueba = 'el pErro de la tiA chOnita'
def Vocales(voc):
    v = 'aeouiAEOUI'
    return [c for c in voc if c in v]#Pedimos que regrese solo los caracteres aeoui, no veo la necesidad de ponerle
#mas condiciones al ciclo for, posiblemente se podria si la variable v no tuviera las vocales en mayusculas, pero no es el caso. 
print(Vocales(cadena_prueba))
print(len(Vocales(cadena_prueba)))

['e', 'E', 'o', 'e', 'a', 'i', 'A', 'O', 'i', 'a']
10


### Ejercicio 8 (2 puntos):

Escribe una clase "calculadora" que regrese el resultado de las operaciones de suma, resta, multiplicacion, division, modulo y potencia. La clase debe contener un método para cada operación.

Los métodos que pudiesen presentar errores como division entre cero, deben ser manejados de tal manera que el programa no imprima ningun error (vimos una clausula para hacer esto).

Adicional a esto, los métodos modulo y potencia deben ser generales, es decir, deben funcionar para cualquier número entero positivo (es un parámetro extra del método). Por ejemplo, la potencia de 2 elevado a 3 es 8, la potencia de 5 elevado a 3 es 125, etc.

Crea una instancia de la clase calculadora y prueba cada uno de los métodos, con valores de tu elección. Para los métodos que pudiesen presentar errores da dos ejemplos, uno totalmente funcional (3/5 por ejemplo) y otro que sea conflictivo pero que no regresa ningún error (4/0 por ejemplo).

In [24]:
#Primero preguntamos dos numeros para trabajar. 
n1 = float(input("Introduce tu primer número: ") )
n2 = float(input("Introduce tu segundo número: ") )
#mensajito de que quieres hacer con esos dos numeros.
opcion = 0
while True:
    print("""
    ¿Que operacion haremos?
    
    1) Sumar los dos números
    2) Restar los dos números
    3) Multiplicar los dos números
    4) Division
    5) Modulo
    6) Potencia
    7) Cambiar los números elegidos
    8) Apagar calculadora
    """)
    opcion = int(input("Elige una opción: ") )     

    if opcion == 1:
        print(" ")
        print("RESULTADO: La suma de",n1,"+",n2,"es igual a",n1+n2)#suma
    elif opcion == 2:
        print(" ")
        print("RESULTADO: La resta de",n1,"-",n2,"es igual a",n1-n2)#resta
    elif opcion == 3:
        print(" ")
        print("RESULTADO: El producto de",n1,"*",n2,"es igual a",n1*n2)#multiplicacion
    elif opcion == 4:
        print(" ")
        print("RESULTADO: La division de",n1,"/",n2,"es igual a",n1/n2) #division pero con enteros
        #Le puse un if para n2 != 0 y que saliera un mensaje de que no se puede dividir entre cero pero no jalaba. 
    elif opcion == 5:
        print(" ")
        print("RESULTADO: El modulo de",n1,"y",n2,"es igual a",n1%n2)#modulo
    elif opcion == 6:
        print(" ")
        print("RESULTADO: La potencia de",n1,"^",n2,"es igual a",n1**n2)#potencia    
    elif opcion == 7:#nuevos numeros
        n1 = float(input("Introduce tu primer número: ") )
        n2 = float(input("Introduce tu segundo número: ") )
    elif opcion == 8:#termimar programa.
        break
    else:#por si apretas un numero no especificado anteriormente
        print("Opción incorrecta")

KeyboardInterrupt: Interrupted by user

### Ejercicio de puntos extras opcional (1 punto):

Supóngamos que tenemos una lista de números enteros y queremos encontrar el número más grande. Por ejemplo, si la lista es `[1, 2, 3, 4, 5]` el número más grande es 5. Crea una función que encuentre el número más grande en una lista de números. La función debe regresar el número más grande y el índice en el que se encuentra. Prueba la función con una lista de números de tu elección.

### Ejercicio de puntos extras opcional (3 puntos):

Supóngamos que tenemos un dado convencional de 6 caras. Crea una función que simule el lanzamiento de un dado. La función debe regresar un número aleatorio entre 1 y 6.

Sabemos que la probabilidad de que salga un número es de $\frac{1}{6}$, es decir, la probabilidad de que salga un 1 es de $\frac{1}{6}$, la probabilidad de que salga un 2 es de $\frac{1}{6}$, etc.

* Prueba la función con 6 lanzamientos, en una celda markdown explica porqué el resultado es dificilmente [1, 2, 3, 4, 5, 6].

* Prueba la función con 1000 lanzamientos, crea una grafica de barras de los resultados (investiga como puedes graficar los resultados en una grafica de barras) y explica a que distribucion se parece y porqué es así.

Hint: Para crear los números aleatorios, puedes usar la función `random.randint()`. O bien, usar el modulo numpy para crearlos. Para graficar los resultados, puedes usar la libreria matplotlib, seaborn, plotly o cualquiera que se te haga más fácil.